repo/maintenancestats/types.go

```diff
@@ -0,0 +1,449 @@
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/kopia/kopia/internal/contentlog"
+)
+
+// RawStats holds the raw data for maintenance statistics
+type RawStats struct {
+	Type string          `json:"type,omitempty"`
+	Raw  json.RawMessage `json:"raw,omitempty"`
+}
+
+// Stats defines the methods for maintenance statistics
+type Stats interface {
+	Type() string
+	Summary() string
+}
+
+// BuildRaw builds a type of Stats into RawStats
+func BuildRaw(stats Stats) (RawStats, error) {
+	bytes, err := json.Marshal(stats)
+	if err != nil {
+		return RawStats{}, err
+	}
+
+	return RawStats{
+		Type: stats.Type(),
+		Raw:  bytes,
+	}, nil
+}
+
+// BuildFromRaw a RawStats into Stats
+func BuildFromRaw(raw RawStats) (Stats, error) {
+	switch raw.Type {
+	case cleanupMarkersStatsType:
+		var cs CleanupMarkersStats
+		if err := json.Unmarshal(raw.Raw, &cs); err != nil {
+			return nil, err
+		}
+
+		return &cs, nil
+	case cleanupSupersededIndexesStatsType:
+		var cs CleanupSupersededIndexesStats
+		if err := json.Unmarshal(raw.Raw, &cs); err != nil {
+			return nil, err
+		}
+
+		return &cs, nil
+	case generateRangeCheckpointStatsType:
+		var gs GenerateRangeCheckpointStats
+		if err := json.Unmarshal(raw.Raw, &gs); err != nil {
+			return nil, err
+		}
+
+		return &gs, nil
+
+	case advanceEpochStatsType:
+		var as AdvanceEpochStats
+		if err := json.Unmarshal(raw.Raw, &as); err != nil {
+			return nil, err
+		}
+
+		return &as, nil
+	case compactSingleEpochStatsType:
+		var cs CompactSingleEpochStats
+		if err := json.Unmarshal(raw.Raw, &cs); err != nil {
+			return nil, err
+		}
+
+		return &cs, nil
+	case compactStatsType:
+		var cs CompactStats
+		if err := json.Unmarshal(raw.Raw, &cs); err != nil {
+			return nil, err
+		}
+
+		return &cs, nil
+	case deleteUnreferencedBlobsStatsType:
+		var ds DeleteUnreferencedBlobsStats
+		if err := json.Unmarshal(raw.Raw, &ds); err != nil {
+			return nil, err
+		}
+
+		return &ds, nil
+	case extendBlobRetentionStatsType:
+		var es ExtendBlobRetentionStats
+		if err := json.Unmarshal(raw.Raw, &es); err != nil {
+			return nil, err
+		}
+
+		return &es, nil
+	case cleanupLogsStatsType:
+		var cs CleanupLogsStats
+		if err := json.Unmarshal(raw.Raw, &cs); err != nil {
+			return nil, err
+		}
+
+		return &cs, nil
+	case rewriteContentsStatsType:
+		var rs RewriteContentsStats
+		if err := json.Unmarshal(raw.Raw, &rs); err != nil {
+			return nil, err
+		}
+
+		return &rs, nil
+	case snapshotGCStatsType:
+		var ss SnapshotGCStats
+		if err := json.Unmarshal(raw.Raw, &ss); err != nil {
+			return nil, err
+		}
+
+		return &ss, nil
+	default:
+		return nil, errors.New("unsupported stats type")
+	}
+}
+
+const cleanupMarkersStatsType = "cleanupMarkersStats"
+
+// CleanupMarkersStats delivers the statistics for CleanupMarkers
+type CleanupMarkersStats struct {
+	EpochMarkers       uint32 `json:"epochMarkers"`
+	DeletionWaterMarks uint32 `json:"deletionWaterMarks"`
+}
+
+// WriteValueTo writes CleanupMarkersStats to JSONWriter
```

Comment:
Simplify this comment, it's redundant and does not provide any additional information that cannot be derived from the the function signature.